File: Companies/Facebook/PartitionEqualSubsetSum.java
package Companies.Facebook;

import java.util.Arrays;

public class PartitionEqualSubsetSum {
    public boolean canPartition(int[] nums) {
        int sum = 0;
        for (int num : nums) {
            sum += num;
        }
        if (sum % 2 != 0) {
            return false;
        }
        Arrays.sort(nums);
        if (nums[nums.length-1] > sum/2) {
            return false;
        }
        return dfs(nums, sum/2, nums.length-1);
    }

    private boolean dfs(int[] nums, int target, int index) {
        if (target == 0) {
            return true;
        }
        for (int i = index; i >= 0; i--) {
            if (nums[i] <= target && dfs(nums, target-nums[i], i-1)) {
                return true;
            }
        }
        return false;
    }

    public boolean canPartitionII(int[] nums) {
        int sum = 0;
        for (int num : nums) {
            sum += num;
        }
        if (sum % 2 != 0) {
            return false;
        }
        int target = sum/2;
        boolean[] dp = new boolean[target+1];
        dp[0] = true;

        for (int num : nums) {
            for (int add = target; add >= num ; add--) {
                dp[add] = dp[add] || dp[add-num];
            }
        }
        return dp[target];
    }
}

File: Companies/VMware/PartitionEqualSubsetSum.cpp
#include <iostream>
#include <vector>
#include <bitset>

using namespace std;

/**
 * num = 5, 000000100001
 * We can achieve 0 and 5 with [ ] and [ 5 ]
 * num = 2, 000010100101
 * We can achieve 0,2,5,7 from [5,2] ==> [ ], [5], [2], [5,2]
 * num = 4, 101011110101
 * We can achieve 0,2,4,5,6,7,11 from [5,2] ==> [ ], [5], [2], [4], [5,2], [2,4], [5,4], [5,2,4]
 */
class PartitionEqualSubsetSum
{
public:
  bool canPartition(vector<int> &nums)
  {
    bitset<10000> bits(1);
    int sum{0}; // accumulate(nums.begin(), nums.end(), 0)
    for (auto num : nums)
    {
      sum += num;
      bits |= bits << num;
    }
    return !(sum % 2) && bits[sum / 2];
  }
};